src/os/newstore/newstore_types.cc

```diff
@@ -240,7 +246,13 @@ void wal_op_t::decode(bufferlist::iterator& p)
   ::decode(fid, p);
   ::decode(offset, p);
   ::decode(length, p);
-  ::decode(data, p);
+  ::decode(data_in_overlay, p);
+  if (!data_in_overlay) {
+    ::decode(data, p);
+  } else {
+    ::decode(nid, p);
+    ::decode(overlays, p);
+  }
```

Comment:
encoding an empty bufferlist is just a (u32)0; no need to do a conditional here (and we can drop data_in_overlay)